src/inline_module.cpp

```diff
@@ -0,0 +1,158 @@
+        if(sins->name() == "@literal")
+        {
+            auto l   = sins->get_literal();
+            copy_ins = m.add_literal(l);
+        }
+        else if(sins->name() == "@param")
+        {
+            auto&& name = any_cast<builtin::param>(sins->get_operator()).parameter;
+            auto s      = sins->get_shape();
+            copy_ins    = m.add_parameter(name, s);
+        }
+        else if(sins->name() == "@outline")
+        {
+            auto s   = sins->get_shape();
+            copy_ins = m.add_outline(s);
+        }
+        else
+        {
+            auto mod_args = sins->module_inputs();
+            auto inputs   = sins->inputs();
+            std::vector<instruction_ref> copy_inputs(inputs.size());
+            std::transform(inputs.begin(), inputs.end(), copy_inputs.begin(), [&](auto i) {
+                return contains(map_ins, i) ? map_ins[i] : i;
+            });
+
+            if(sins->name() == "@return")
+            {
+                mod_outputs = copy_inputs;
+                break;
+            }
+
+            if(mod_args.empty())
+            {
+                copy_ins = m.insert_instruction(ins, sins->get_operator(), copy_inputs);
+            }
+            else
+            {
+                copy_ins = m.insert_instruction(ins, sins->get_operator(), copy_inputs, mod_args);
+            }
+        }
+        map_ins[sins] = copy_ins;
+        mod_outputs   = {copy_ins};
+    }
+
+    auto ins_outputs = ins->outputs();
+    assert(mod_outputs.size() >= ins_outputs.size());
+    for(const auto& out : ins_outputs)
+    {
+        auto val = out->get_operator().to_value();
+        assert(val.contains("index"));
+        auto index = val.at("index").to<std::size_t>();
+        m.replace_instruction(out, mod_outputs.at(index));
+    }
+}
+
+void inline_module::apply(module& m) const
+{
+    for(auto ins : iterator_for(m))
+    {
+        if(ins->name() != "if")
+            continue;
+
+        auto arg_cond          = ins->inputs().front()->eval();
+        const auto& mod_inputs = ins->module_inputs();
+        // condition is not constant, but both subgraph outputs
+        // are constant, so we can replace each subgraph with
+        // a literal
+        if(arg_cond.empty())
+        {
+            std::vector<argument> arg_outs;
+            for(const auto& mod : mod_inputs)
+            {
+                auto last = std::prev(mod->end());
+                std::vector<instruction_ref> mod_outputs;
+                if(last->name() == "@return")
+                {
+                    mod_outputs = last->inputs();
+                }
+                else
+                {
+                    // no return instruction, last instruction is output
+                    mod_outputs.push_back(last);
+                }
+
+                // only one output is considered for now
+                for(const auto& out : mod_outputs)
+                {
+                    auto mod_out = out->eval();
```

Comment:
Actually, I think it would be simpler to leave const propagation to the const propagate pass and then just take of inlining when the condition is known.

Reply:
Actually, I am thinking about extending the if operator as the following:
1) With module input, we have the current implementation. 
2) No module inputs but with 3 input arguments (cond, then_arg, else_arg). Then if cond is true, return "then_arg"; if cond is false, return else_arg. 
With this extension, this optimization is much simpler, and the optimization for "find_where_op" can be further optimized.

Any opinion?

Reply:
I just created another branch (extend the if, further optimize for here and the find_where_op) to implemented what I just mentioned. So for this PR, I would prefer keep this.

Reply:
just removed this optimization